
// DRP: Distrend protocol.

// Type of request.
enum RequestType {
    WELCOME = 1;
    COPY_IN = 2;
    EXECUTE = 3;
    COPY_OUT = 4;
}

message WelcomeRequest {
    // Nothing.
}

message CopyInRequest {
    optional string pathname = 1;
    optional bytes content = 2;
}

message ExecuteRequest {
    optional string executable = 1;

    // Arguments don't include the executable.
    repeated string arguments = 2;
}

message CopyOutRequest {
    optional string pathname = 1;
}

// Request from controller to worker.
message Request {
    // Client ID can be missing for welcome requests.
    optional string client_id = 1;
    optional RequestType request_type = 2;

    optional WelcomeRequest welcome_request = 10;
    optional CopyInRequest copy_in_request = 11;
    optional ExecuteRequest execute_request = 12;
    optional CopyOutRequest copy_out_request = 13;
}

message WelcomeResponse {
    // Nothing.
}

message CopyInResponse {
    optional bool success = 1;
}

message ExecuteResponse {
    optional int32 status = 1;
}

message CopyOutResponse {
    // If missing, then file wasn't found or wasn't readable.
    optional bytes content = 2;
}

message Response {
    optional string client_id = 1;
    optional RequestType request_type = 2;

    optional WelcomeResponse welcome_response = 10;
    optional CopyInResponse copy_in_response = 11;
    optional ExecuteResponse execute_response = 12;
    optional CopyOutResponse copy_out_response = 13;
}
